cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -131,23 +130,163 @@ __device__ __host__ double attractive_grad(double dist_squared,
  * randomly.
  */
 template <typename T, int TPB_X>
-__global__ void optimize_batch_kernel(
+__global__ void optimize_batch_kernel_on_shared_mem(
   T *head_embedding, int head_n, T *tail_embedding, int tail_n, const int *head,
   const int *tail, int nnz, T *epochs_per_sample, int n_vertices,
   bool move_other, T *epochs_per_negative_sample,
   T *epoch_of_next_negative_sample, T *epoch_of_next_sample, double alpha,
   int epoch, double gamma, uint64_t seed, UMAPParams params) {
+  extern __shared__ T embedding_updates[];
+
+  int row = (blockIdx.x * TPB_X) + threadIdx.x;
+  if (row < nnz) {
+    /**
+     * Positive sample stage (attractive forces)
+     */
+
+    if (epoch_of_next_sample[row] <= epoch) {
+      int j = head[row];
+      int k = tail[row];
+
+      T *current = head_embedding + (j * params.n_components);
+      T *other = tail_embedding + (k * params.n_components);
+      T *current_buffer, *other_buffer;
```

Comment:
The current approach is making each thread access its components from contiguous locations. This leads to shared mem bank conflicts. Better approach is to transpose it. Meaning, for the below if-else block the value of these buffers will be:
```cpp
current_buffer = embedding_updates + threadIdx.x;
other_buffer = embedding_updates + TPB_X * params.n_components + threadIdx.x;
```

Then, accessing the "thread local" components will be:
```cpp
for (int d = 0; d < params.n_components; ++i) {
  auto current_val = current_buffer[d * TPB_X];
  auto other_val = other_buffer[d * TPB_X];
  ...
}
```

Reply:
Ah ok, never considered that